File: 20230107_Saturday/20230107.rb
# Leetcode: 134. Gas Station.
# https://leetcode.com/problems/gas-station/
# = = = = = = = = = = = = = =
# Accpeted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 137 ms, faster than 91.67% of Ruby online submissions for Gas Station.
# Memory Usage: 217.6 MB, less than 16.67% of Ruby online submissions for Gas Station.
# 2023.01.07 Daily Challenge.
# @param {Integer[]} gas
# @param {Integer[]} cost
# @return {Integer}
def can_complete_circuit(gas, cost)
  # Based on:
  # https://leetcode.com/problems/gas-station/discuss/42568/Share-some-of-my-ideas.
  tank, all_way,all_gas,start = 0,0,0,0
  gas.size.times do |i|
    all_way += cost[i]
    all_gas += gas[i]
    if tank + gas[i] < cost[i]
      start = i+1 # Here was an error. Fixed.
      tank = 0
    else
      tank += gas[i] - cost[i]
    end
  end
  return -1 if all_gas < all_way
  return start
end

File: 20230107_Saturday/20230107.go
// Leetcode: 134. Gas Station.
// https://leetcode.com/problems/gas-station/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 82 ms, faster than 87.77% of Go online submissions for Gas Station.
// Memory Usage: 9.1 MB, less than 99.28% of Go online submissions for Gas Station.
// 2023.01.07 Daily Challenge.
func canCompleteCircuit(gas []int, cost []int) int {
  // Based on:
  // https://leetcode.com/problems/gas-station/discuss/42568/Share-some-of-my-ideas.
  start,allGas,allCost,tank := 0,0,0,0
  for i,g := range gas {
    allGas += g
    allCost += cost[i]
    if tank + g < cost[i] {
      tank = 0
      start = i + 1
    } else {
      tank += (g - cost[i])
    }
  }
  if allGas < allCost { return -1 }
  return start
}
